### PR TITLE
Use lists for both the existing and the new instances.

### PR DESCRIPTION
**Pull request by simonvanderveldt**
This allows us to compare the 2 lists and first start the instances that don't exist yet, then cycle the instances that already exist and then stop the instances that no longer should exist.
Also added is the removal + a warning message of a non-instanced unit and non-matching instances for this service.

I'm not 100% happy with the regex I had to add to `FleetService`, whereas the other ones all live in `FleetHelper`. Maybe it makes sense to move the matching of units that aren't supported there as well?

@amochtar can you review/feedback?

Sample output with debugging:

```
# Units before
core@core-01 /tmp $ fleetctl list-units
UNIT            MACHINE             ACTIVE  SUB
crap.service        78cd7666.../172.17.8.103    active  running
crap@10.service     1e5729bc.../172.17.8.101    active  running
crap@2.service      1e5729bc.../172.17.8.101    active  running
crap@3.service      78cd7666.../172.17.8.103    active  running
crap@abc.service    dd5358c0.../172.17.8.102    active  running

(.venv) ~/s/s/fleet-service:use-lists-for-new-and-existing-instances ❯ ./fs -v DEBUG create --count 2 crap crap.service
Creating service crap
Desired instance count: 2
Desired instances: [u'crap@1.service', u'crap@2.service']
Existing instance count: 3
Existing instances: [u'crap@10.service', u'crap@2.service', u'crap@3.service']
warning: Destroying non-instance unit crap.service
debug: Destroying unit: crap.service
debug: Waiting for unit crap.service to reach state None
debug: crap.service state: None
warning: Destroying non-matching instance units [u'crap@abc.service']
debug: Destroying unit: crap@abc.service
debug: Waiting for unit crap@abc.service to reach state None
debug: crap@abc.service state: None
Destroying old template crap@.service
debug: Destroying unit: crap@.service
debug: Waiting for unit crap@.service to reach state None
debug: crap@.service state: None
Submitting new template crap@.service
debug: Creating new unit: crap@.service with desired state inactive
debug: Waiting for unit crap@.service to reach state inactive
debug: crap@.service state: inactive
Creating new instances: set([u'crap@1.service'])
Creating instance crap@1.service
debug: Creating new unit: crap@1.service with desired state launched
debug: Waiting for unit crap@1.service to reach state launched
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: launched
debug: Waiting for unit crap@1.service to reach SystemD state active
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: active
Updating existing instances: set([u'crap@2.service'])
Destroying instance crap@2.service
debug: Destroying unit: crap@2.service
debug: Waiting for unit crap@2.service to reach state None
debug: crap@2.service state: None
debug: Waiting for unit crap@2.service to reach SystemD state None
debug: crap@2.service SystemD state: active
debug: crap@2.service SystemD state: active
debug: crap@2.service SystemD state: active
debug: crap@2.service SystemD state: active
debug: crap@2.service SystemD state: None
Creating instance crap@2.service
debug: Creating new unit: crap@2.service with desired state launched
debug: Waiting for unit crap@2.service to reach state launched
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: launched
debug: Waiting for unit crap@2.service to reach SystemD state active
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: active
Destroying existing instances: set([u'crap@3.service', u'crap@10.service'])
Destroying instance crap@3.service
debug: Destroying unit: crap@3.service
debug: Waiting for unit crap@3.service to reach state None
debug: crap@3.service state: None
debug: Waiting for unit crap@3.service to reach SystemD state None
debug: crap@3.service SystemD state: active
debug: crap@3.service SystemD state: active
debug: crap@3.service SystemD state: active
debug: crap@3.service SystemD state: None
Destroying instance crap@10.service
debug: Destroying unit: crap@10.service
debug: Waiting for unit crap@10.service to reach state None
debug: crap@10.service state: None
debug: Waiting for unit crap@10.service to reach SystemD state None
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: None

# Units after
core@core-01 /tmp $ fleetctl list-units
UNIT        MACHINE             ACTIVE  SUB
crap@1.service  dd5358c0.../172.17.8.102    active  running
crap@2.service  1e5729bc.../172.17.8.101    active  running
```

### CODE REVIEW

File: fleet_helper.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def get_unit_instances(self, units, service_name):
         unit_instance_pattern = re.compile(r"^" + re.escape(service_name) + r"@(\d+)\.service")
         for unit in units:
             if unit_instance_pattern.match(unit['name']):
-                unit_instances.append(unit)
+                unit_instances.append(unit.name)
```
Comment (amochtar) [inline fleet_helper.py:65]: You can shorten this to something like:

```
unit_instance_pattern = re.compile(r"^" + re.escape(service_name) + r"@(\d+)\.service")
return [ unit.name for unit in units if unit_instance_pattern.match(unit.name) ]
```

File: fleet_service.py [hunk 2]
```diff
@@ -22,43 +23,70 @@ def create_service(self, service_name, unit_file, count=3):
...
+        for instance in instances_to_create:
+            self.create_instance(instance, instance_unit)
+
+        # Update instances that already exist
+        instances_to_update = set(existing_service_instances) & set(new_service_instances)
+        if len(instances_to_update) > 0:
+            self.logger.info('Updating existing instances: ' + str(instances_to_update))
+        for instance in instances_to_update:
+            self.destroy_instance(instance)
+            self.create_instance(instance, instance_unit)
```
Comment (amochtar) [inline fleet_service.py]: Add replace function to wrap destroy and create

Comment (simonvanderveldt) [inline fleet_service.py]: I've added a `destroy_and_create` method

**Comment by simonvanderveldt** [comment]
I've moved some parts around so that `fleet_helper` doesn't have anything to do with the service concept and that it's no longer necessary to pass from the outside the list of units to some of it's methods.

Code as per this PR works for what we want, though I do think a small refactor is in order to get rid of the inheritance. We should probably introduce a `Service` class or something like that that manages/owns a specific service,

Sample output:

```
(.venv) ~/s/s/fleet-service:use-lists-for-new-and-existing-instances ✹ ❯ ./fs -v DEBUG create --count 5 crap crap@.service
Creating service crap
Desired instance count: 5
Desired instances: [u'crap@1.service', u'crap@2.service', u'crap@3.service', u'crap@4.service', u'crap@5.service']
Existing instance count: 8
Existing instances: [u'crap@10.service', u'crap@15.service', u'crap@2.service', u'crap@3.service', u'crap@4.service', u'crap@6.service', u'crap@abc.service', u'crap@def.service']
Existing instances managed by us: [u'crap@10.service', u'crap@15.service', u'crap@2.service', u'crap@3.service', u'crap@4.service', u'crap@6.service']
warning: Destroying non-instance unit crap.service
debug: Destroying unit: crap.service
debug: Waiting for unit crap.service to reach state None
debug: crap.service state: None
debug: Waiting for unit crap.service to reach SystemD state None
debug: crap.service SystemD state: active
debug: crap.service SystemD state: active
debug: crap.service SystemD state: active
debug: crap.service SystemD state: active
debug: crap.service SystemD state: active
debug: crap.service SystemD state: active
debug: crap.service SystemD state: active
debug: crap.service SystemD state: active
debug: crap.service SystemD state: None
warning: Destroying non-matching instance units [u'crap@abc.service', u'crap@def.service']
debug: Destroying unit: crap@abc.service
debug: Waiting for unit crap@abc.service to reach state None
debug: crap@abc.service state: None
debug: Waiting for unit crap@abc.service to reach SystemD state None
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: active
debug: crap@abc.service SystemD state: None
debug: Destroying unit: crap@def.service
debug: Waiting for unit crap@def.service to reach state None
debug: crap@def.service state: None
debug: Waiting for unit crap@def.service to reach SystemD state None
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: active
debug: crap@def.service SystemD state: None
Destroying old template crap@.service
debug: Destroying unit: crap@.service
debug: Waiting for unit crap@.service to reach state None
debug: crap@.service state: None
debug: Waiting for unit crap@.service to reach SystemD state None
debug: crap@.service SystemD state: None
Submitting new template crap@.service
debug: Creating new unit: crap@.service with desired state inactive
debug: Waiting for unit crap@.service to reach state inactive
debug: crap@.service state: inactive
Creating new instances: [u'crap@1.service', u'crap@5.service']
debug: Creating new unit: crap@1.service with desired state launched
debug: Waiting for unit crap@1.service to reach state launched
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: inactive
debug: crap@1.service state: launched
debug: Waiting for unit crap@1.service to reach SystemD state active
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: activating
debug: crap@1.service SystemD state: active
debug: Creating new unit: crap@5.service with desired state launched
debug: Waiting for unit crap@5.service to reach state launched
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: inactive
debug: crap@5.service state: launched
debug: Waiting for unit crap@5.service to reach SystemD state active
debug: crap@5.service SystemD state: activating
debug: crap@5.service SystemD state: activating
debug: crap@5.service SystemD state: activating
debug: crap@5.service SystemD state: activating
debug: crap@5.service SystemD state: activating
debug: crap@5.service SystemD state: activating
debug: crap@5.service SystemD state: active
Updating existing instances: [u'crap@2.service', u'crap@3.service', u'crap@4.service']
debug: Destroying unit: crap@2.service
debug: Waiting for unit crap@2.service to reach state None
debug: crap@2.service state: None
debug: Waiting for unit crap@2.service to reach SystemD state None
debug: crap@2.service SystemD state: active
debug: crap@2.service SystemD state: active
debug: crap@2.service SystemD state: active
debug: crap@2.service SystemD state: active
debug: crap@2.service SystemD state: None
debug: Creating new unit: crap@2.service with desired state launched
debug: Waiting for unit crap@2.service to reach state launched
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: inactive
debug: crap@2.service state: launched
debug: Waiting for unit crap@2.service to reach SystemD state active
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: activating
debug: crap@2.service SystemD state: active
debug: Destroying unit: crap@3.service
debug: Waiting for unit crap@3.service to reach state None
debug: crap@3.service state: None
debug: Waiting for unit crap@3.service to reach SystemD state None
debug: crap@3.service SystemD state: active
debug: crap@3.service SystemD state: active
debug: crap@3.service SystemD state: active
debug: crap@3.service SystemD state: active
debug: crap@3.service SystemD state: None
debug: Creating new unit: crap@3.service with desired state launched
debug: Waiting for unit crap@3.service to reach state launched
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: inactive
debug: crap@3.service state: launched
debug: Waiting for unit crap@3.service to reach SystemD state active
debug: crap@3.service SystemD state: activating
debug: crap@3.service SystemD state: activating
debug: crap@3.service SystemD state: activating
debug: crap@3.service SystemD state: activating
debug: crap@3.service SystemD state: activating
debug: crap@3.service SystemD state: activating
debug: crap@3.service SystemD state: active
debug: Destroying unit: crap@4.service
debug: Waiting for unit crap@4.service to reach state None
debug: crap@4.service state: None
debug: Waiting for unit crap@4.service to reach SystemD state None
debug: crap@4.service SystemD state: active
debug: crap@4.service SystemD state: active
debug: crap@4.service SystemD state: active
debug: crap@4.service SystemD state: active
debug: crap@4.service SystemD state: None
debug: Creating new unit: crap@4.service with desired state launched
debug: Waiting for unit crap@4.service to reach state launched
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: inactive
debug: crap@4.service state: launched
debug: Waiting for unit crap@4.service to reach SystemD state active
debug: crap@4.service SystemD state: activating
debug: crap@4.service SystemD state: activating
debug: crap@4.service SystemD state: activating
debug: crap@4.service SystemD state: activating
debug: crap@4.service SystemD state: activating
debug: crap@4.service SystemD state: active
Destroying existing instances: [u'crap@10.service', u'crap@15.service', u'crap@6.service']
debug: Destroying unit: crap@10.service
debug: Waiting for unit crap@10.service to reach state None
debug: crap@10.service state: None
debug: Waiting for unit crap@10.service to reach SystemD state None
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: active
debug: crap@10.service SystemD state: None
debug: Destroying unit: crap@15.service
debug: Waiting for unit crap@15.service to reach state None
debug: crap@15.service state: None
debug: Waiting for unit crap@15.service to reach SystemD state None
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: active
debug: crap@15.service SystemD state: None
debug: Destroying unit: crap@6.service
debug: Waiting for unit crap@6.service to reach state None
debug: crap@6.service state: None
debug: Waiting for unit crap@6.service to reach SystemD state None
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: active
debug: crap@6.service SystemD state: None
```

**Comment by amochtar** [comment]
> Code as per this PR works for what we want, though I do think a small refactor is in order to get rid of the inheritance. We should probably introduce a Service class or something like that that manages/owns a specific service

Agree, that could clean things up a bit. But this looks fine for now :)
